File: src/main/kotlin/grokkingfp/ch03_2.java
package grokkingfp;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ch03_2 {

    public static String abbreviate(String before) {
        return "";
    }

    public static void main(String[] args) {
        List<String> appleBook = Arrays.asList("Apple", "Book");
        List<String> appleBookMango = new ArrayList<>(appleBook);
        appleBookMango.add("Mango");

        System.out.println(appleBook.size());
        System.out.println(appleBookMango.size());

        /* *******************/
        System.out.println(abbreviate("Alonzo Church")); // "A. Church"
    }
}

File: src/main/kotlin/grokkingfp/ch03_2.kt
package grokkingfp

fun abbreviate(name: String): String {
    val initial = name.substring(0, 1)
    val separator = name.indexOf(' ')
    val lastName = name.substring(separator + 1)
    return "$initial. $lastName"
}

fun firstTwo(list: List<String>): List<String> = list.slice(0..2)

fun main() {
    /**
     * TODO - check immutability
     */
    val appleBook = listOf("Apple", "Book")
    val appleBookMango = appleBook.plus("Mango")

    println(appleBook.size) // 2
    println(appleBookMango.size) // 3

}
